sunbeam-python/sunbeam/core/watcher.py

```diff
@@ -0,0 +1,175 @@
+SLEEP_INTERVAL = 5
+ENABLE_MAINTENANCE_AUDIT_TEMPLATE_NAME = "Sunbeam Cluster Maintaining Template"
+ENABLE_MAINTENANCE_STRATEGY_NAME = "host_maintenance"
+ENABLE_MAINTENANCE_GOAL_NAME = "cluster_maintaining"
+
+WORKLOAD_BALANCING_GOAL_NAME = "workload_balancing"
+WORKLOAD_BALANCING_STRATEGY_NAME = "workload_stabilization"
+WORKLOAD_BALANCING_AUDIT_TEMPLATE_NAME = "Sunbeam Cluster Workload Balancing Template"
+
+
+def get_watcher_client(deployment: Deployment) -> watcher_client.Client:
+    region = read_config(deployment.get_client(), REGION_CONFIG_KEY)["region"]
+    conn = get_admin_connection(
+        jhelper=JujuHelper(deployment.get_connected_controller())
+    )
+
+    watcher_endpoint = conn.session.get_endpoint(
+        service_type="infra-optim",
+        region_name=region,
+    )
+    return watcher_client.Client(session=conn.session, endpoint=watcher_endpoint)
+
+
+def _create_host_maintenance_audit_template(
+    client: watcher_client.Client,
+) -> watcher.AuditTemplate:
+    template = client.audit_template.create(
+        name=ENABLE_MAINTENANCE_AUDIT_TEMPLATE_NAME,
+        description="Audit template for cluster maintaining",
+        goal=ENABLE_MAINTENANCE_GOAL_NAME,
+        strategy=ENABLE_MAINTENANCE_STRATEGY_NAME,
+    )
+    return template
+
+
+def _create_workload_balancing_audit_template(
+    client: watcher_client.Client,
+) -> watcher.AuditTemplate:
+    template = client.audit_template.create(
+        name=WORKLOAD_BALANCING_AUDIT_TEMPLATE_NAME,
+        description="Audit template for workload balancing",
+        goal=WORKLOAD_BALANCING_GOAL_NAME,
+        strategy=WORKLOAD_BALANCING_STRATEGY_NAME,
+    )
+    return template
+
+
+def get_enable_maintenance_audit_template(
+    client: watcher_client.Client,
+) -> watcher.AuditTemplate:
+    try:
+        template = client.audit_template.get(ENABLE_MAINTENANCE_AUDIT_TEMPLATE_NAME)
+    except NotFound:
+        template = _create_host_maintenance_audit_template(client=client)
+    return template
+
+
+def get_workload_balancing_audit_template(
+    client: watcher_client.Client,
+) -> watcher.AuditTemplate:
+    try:
+        template = client.audit_template.get(WORKLOAD_BALANCING_AUDIT_TEMPLATE_NAME)
+    except NotFound:
+        template = _create_workload_balancing_audit_template(client=client)
+    return template
+
+
+@timeout_decorator.timeout(TIMEOUT)
+def create_audit(
+    client: watcher_client.Client,
+    template: watcher.AuditTemplate,
+    audit_type: str = "ONESHOT",
+    parameters: dict[str, Any] = {},
+) -> watcher.Audit:
+    audit = client.audit.create(
+        audit_template_uuid=template.uuid,
+        audit_type=audit_type,
+        parameters=parameters,
+    )
+    while True:
+        audit_details = client.audit.get(audit.uuid)
+        if audit_details.state in ["SUCCEEDED", "FAILED"]:
+            break
+        time.sleep(SLEEP_INTERVAL)
+    if audit_details.state == "SUCCEEDED":
```

Comment:
The way this is set up with a timeout with a loop condition could be refactored to use tenacity retry quite elegantly I think. For example:

```suggestion
@timeout_decorator.timeout(TIMEOUT)
@tenacity.retry(wait=wait_fixed(SLEEP_INTERVAL), stop=tenacity.stop_after_delay(TIMEOUT))
def create_audit(
    client: watcher_client.Client,
    template: watcher.AuditTemplate,
    audit_type: str = "ONESHOT",
    parameters: dict[str, Any] = {},
) -> watcher.Audit:
    audit = client.audit.create(
        audit_template_uuid=template.uuid,
        audit_type=audit_type,
        parameters=parameters,
    )
    audit_details = client.audit.get(audit.uuid)
    if audit_details.state not in ["SUCCEEDED", "FAILED"]:
        raise Exception(...)
    if audit_details.state == "SUCCEEDED":
```

(unless there's something I'm missing that caused you to need to use the timeout decorator?)

Reply:
Nice suggestion!

Reply:
Hmm good point.

An idea: it could be split into two functions:

```python
def create_audit(...):
  audit = client.audit.create(...)
  return _inner_create_audit(audit, ...)

@retry(...)
def _inner_create_audit(audit, ...):
    ...
```

Reply:
Ah I just read your change - nice refactoring :ok_hand:

Reply:
Updated.